language-plutus-core/src/Language/PlutusCore/Name.hs

```diff
@@ -78,19 +81,31 @@ newtype Unique = Unique { unUnique :: Int }
     deriving (Eq, Show, Ord, Lift)
     deriving newtype (NFData)
 
+-- | The unique of a type-level name.
+newtype TypeUnique = TypeUnique
+    { unTypeUnique :: Unique
+    }
+
+-- | The unique of a term-level name.
+newtype TermUnique = TermUnique
+    { unTermUnique :: Unique
+    }
+
 -- | Types which have a 'Unique' attached to them, mostly names.
-class HasUnique a where
-    unique :: Lens' a Unique
+class Coercible Unique unique => HasUnique a unique | a -> unique where
+    unique :: Lens' a unique
 
-instance HasUnique (Name a) where
-    unique = lens g s where
-        g = nameUnique
-        s n u = n{nameUnique=u}
+    -- The default instance for newtypes.
+    default unique :: (Coercible (Unwrap a) a, HasUnique (Unwrap a) unique', Coercible unique' unique)
```

Comment:
TBH I'd be happy just copying this definition into the few places where we use this, rather than this somewhat complex default version.

Reply:
Well, then I'd like to make `newtypeUnique` with the same signature and the implementation, because why copy-paste when you can just call a function.

Reply:
Removed the default implementation.